### PR TITLE
Kubeadm: Refactor renewal package

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
the code implemented for managing renewal is getting more and more complex after detecting external CA and renewing kubeConfig. This PR refactors this code concentrating all the complexity into one package and making more transparent all the renewal internals to its consumers.

This refactor fixes also an inconsistency between `kubeadm alpha certs renew  --csr-only` and `kubeadm alpha certs renew` and sets the ground for implementing certs expiration checks in a cleaner and less invasive way

**Which issue(s) this PR fixes**:
Rif https://github.com/kubernetes/kubeadm/issues/1361
Rif https://github.com/kubernetes/kubeadm/issues/1540

**Does this PR introduce a user-facing change?**:
```release-note
kubeadm: kubeadm alpha certs renew  --csr-only now reads the current certificates as the authoritative source for certificates attributes (same as kubeadm alpha certs renew)
```
/sig cluster-lifecycle
/area kubeadm
/priority important-soon
@kubernetes/sig-cluster-lifecycle-pr-reviews

/assign @rosti 
/assign @timothysc

### CODE REVIEW

File: cmd/kubeadm/app/phases/certs/renewal/readwriter.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/renewal/readwriter.go]: 2019

File: cmd/kubeadm/app/phases/certs/renewal/readwriter.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+// Write a certificate to files in the K8s pki managed by kubeadm
+func (rw *pkiCertificateReadWriter) Write(newCert *x509.Certificate, newKey crypto.Signer) error {
+	if err := pkiutil.WriteCertAndKey(rw.certificateDir, rw.baseName, newCert, newKey); err != nil {
+		return errors.Wrapf(err, "failed to write new certificate %s", rw.baseName)
+	}
+	return nil
+}
+
+// kubeConfigReadWriter defines a certificateReadWriter for certificate files
+// embedded in the kubeConfig files managed managed by kubeadm, and more specifically
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/renewal/readwriter.go]: double `managed`

File: cmd/kubeadm/app/phases/certs/renewal/readwriter.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+	rw.kubeConfig = kubeConfig
+
+	return certs[0], nil
+}
+
+// Write a certificate embedded in kubeConfig file managed by kubeadm
+// Please note that all the other attribute of the kubeConfig file are preserved, but this
+// requires to call Read before Write
+func (rw *kubeConfigReadWriter) Write(newCert *x509.Certificate, newKey crypto.Signer) error {
+	//
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/renewal/readwriter.go]: missing comment?

File: cmd/kubeadm/app/phases/certs/renewal/readwriter_test.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/renewal/readwriter_test.go]: 2019

File: cmd/kubeadm/app/phases/certs/renewal/readwriter_test.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+	// Reads back the new certificate
+	readCert, err = pkiReadWriter.Read()
+	if err != nil {
+		t.Fatalf("couldn't read new certificate: %v", err)
+	}
+
+	// Check if the new certificate read from disk is equal to the original one
+	if !newCert.Equal(readCert) {
+		t.Errorf("read cert does not match with expected new cert")
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/renewal/readwriter_test.go]: `t.Error()`

File: cmd/kubeadm/app/phases/certs/renewal/readwriter_test.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+	// Reads back the new certificate embedded in a kubeconfig writer
+	readCert, err = kubeconfigReadWriter.Read()
+	if err != nil {
+		t.Fatalf("couldn't read new embedded  certificate: %v", err)
+	}
+
+	// Check if the new certificate read from disk is equal to the original one
+	if !newCert.Equal(readCert) {
+		t.Errorf("read cert does not match with expected new cert")
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/renewal/readwriter_test.go]: `t.Error`

File: cmd/kubeadm/app/phases/certs/renewal/readwriter_test.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+// writeTestCertificate is a utility for creating a test certificate
+func writeTestCertificate(t *testing.T, dir, name string, caCert *x509.Certificate, caKey crypto.Signer) *x509.Certificate {
+	cert, key, err := pkiutil.NewCertAndKey(caCert, caKey, testCertCfg)
+	if err != nil {
+		t.Fatalf("couldn't generate certificate: %v", err)
+	}
+
+	if err := pkiutil.WriteCertAndKey(dir, name, cert, key); err != nil {
+		t.Fatalf("couldn't write out certificate")
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/renewal/readwriter_test.go]: `t.Fatal` or add e.g. `dir` or `name`

**Review by neolit123 — COMMENTED** [review]
thanks @fabriziopandini 
that's a pretty big refactor and the new manager logic seems valid to me.

i think we should start allowing separate commits per PR, to make the review easier.

**Comment by fabriziopandini** [comment]
@neolit123 thanks for the review!
All comment are addressed

**Review by rosti — COMMENTED** [review]
Thanks @fabriziopandini !
Looks way better than before (especially for `renewCertsByComponent`)!
Found a tiny nit at the first pass, but other than that, I'll try to do another pass tomorrow as the change is large.

File: cmd/kubeadm/app/cmd/alpha/certs.go [hunk 8]
```diff
@@ -97,45 +101,27 @@ func getRenewSubCommands(kdir string) []*cobra.Command {
...
 	}
 	// Default values for the cobra help text
-	kubeadmscheme.Scheme.Default(&cfg.cfg)
+	kubeadmscheme.Scheme.Default(&flags.cfg)
 
-	certTree, err := certsphase.GetDefaultCertList().AsMap().CertTree()
-	kubeadmutil.CheckErr(err)
+	// Get a renewal manager for a generic Cluster configuration, that is used only for getting
+	// the list of certificates for building subcommands
+	rm, _ := renewal.NewManager(&kubeadmapi.ClusterConfiguration{}, "")
```
[the remark this answers is not in the document]
Comment (rosti) [reply cmd/kubeadm/app/cmd/alpha/certs.go]: Actually, the error that is returned by `NewManager` is not looked at anywhere.

[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/alpha/certs.go]: @rosti, good catch!
I somehow forgot to fix them
Now it is addressed

**Comment by neolit123** [comment]
/lgtm

LGTM-ing here to unblock the work.
retrospective reviews are fine, but we need to move forward.

**Comment by fabriziopandini** [comment]
Uhmm.. test seems stuck
/retest

**Pull request merged**